# GROUPING DATA & AGGREGATING VARIABLES (part 2)


### This script contains the following:
#### 1. Importing data and libraries
#### 2. Aggregate mean of order_number by department_id
#### 3. Analyze the results
#### 4. Loyalty flag
#### 5. Basic statistics of product prices
#### 6. Spending flag
#### 7. Order frequency flag

# ---------------------------------------------------------------------------------------------------------------

## 1. Importing libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 1.1. Importing data

In [4]:
# We create a path to add a shortcut to our files location
path = r'C:\Users\javis\OneDrive\Documentos\Career Foundry\Data Immersion\Python\Instacart Basket Analysis'

In [5]:
# We then give Python the instruction of reading the file located in that path
ords_prods_merge = pd.read_pickle(os.path.join(path, '2. Data', '2.2. Prepared Data', 'orders_products_merged_new_variables.pkl'))

## 2. Find aggregate mean of order_number by department_id

In [6]:
# To calculate the average number of orders per department ID
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## 3. Analyze the results

### When comparing the results between a subset and the entire dataframe, we notice that the averages of orders for all of the departments change. For example, the department 17 had an average of 11.29 in the subset but increases to 15.69 when using the whole dataframe as a reference.

## 4. Create a loyalty flag

In [7]:
# Firstly we create a new column, group by criteria and apply the function
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# We check the output 
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_from_last_order  first_order  product_id  add_to_cart_order  \
0                    NaN         True         196                  1   
1                   15.0        False         196                  1   
2                   21.0        False         196                  1   
3                   29.0        False         196                  1   
4                   28.0        False         196                  1   
5                   19.0        False         196                  1   
6                   20.0        False         196                  1   
7                   14.0        False         196                  2   
8                    0.0        False         196                  4   
9                   30.0        False         196                  1   
10                   7.0        False         196                  2   
11                   8.0        False         196                  6   
12                   7.0        False         196                  1   
13                   7.0        False         196                  1   
14                  14.0        False         196                  1   

    reordered _merge product_name  aisle_id  department_id  prices  \
0           0   both         Soda        77              7     9.0   
1           1   both         Soda        77              7     9.0   
2           1   both         Soda        77              7     9.0   
3           1   both         Soda        77              7     9.0   
4           1   both         Soda        77              7     9.0   
5           1   both         Soda        77              7     9.0   
6           1   both         Soda        77              7     9.0   
7           1   both         Soda        77              7     9.0   
8           1   both         Soda        77              7     9.0   
9           1   both         Soda        77              7     9.0   
10          0   both         Soda        77              7     9.0   
11          1   both         Soda        77              7     9.0   
12          1   both         Soda        77              7     9.0   
13          1   both         Soda        77              7     9.0   
14          1   both         Soda        77              7     9.0   

      price_range_loc     busiest_day      Busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy       Normal_days       Most busy hours   
1   Mid-range product  Regularly busy  Two_slowest_days    Average busy hours   
2   Mid-range product  Regularly busy  Two_slowest_days       Most busy hours   
3   Mid-range product      Least busy  Two_slowest_days    Average busy hours   
4   Mid-range product      Least busy  Two_slowest_days       Most busy hour

In [9]:
# With the new column ready, we have to create a flag to denominate the loyalty of the customers
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [11]:
# We do the same for regular customers
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [12]:
# And for new customers
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [13]:
# As a good practice, we check the frequency 
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

 ## 5. Check the basic statistics of the product prices for each loyalty category

In [14]:
# We are going to calculate average, minimum and maximum of the prices for each category of customer
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### As we can see, and against my expectations, it is the new customers who spent more in each product as an average, followed by the regular customers ,being the loyal customers the ones that spends the less in average. That might be due to the fact that they know about offers or they get discount, e.g. 

## 6. Create a spending flag for each user based on the average price across all their orders

In [15]:
# We create a new column, group by criteria and apply the function
ords_prods_merge['average_spending'] = ords_prods_merge.groupby('user_id').agg({'prices':['mean']})

In [16]:
# CHeck the ouput 
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_from_last_order  first_order  product_id  add_to_cart_order  \
0                   NaN         True         196                  1   
1                  15.0        False         196                  1   
2                  21.0        False         196                  1   
3                  29.0        False         196                  1   
4                  28.0        False         196                  1   

   reordered  ... aisle_id department_id  prices    price_range_loc  \
0          0  ...       77             7     9.0  Mid-range product   
1          1  ...       77             7     9.0  Mid-range product   
2          1  ...       77             7     9.0  Mid-range product   
3          1  ...       77             7     9.0  Mid-range product   
4          1  ...       77             7     9.0  Mid-range product   

      busiest_day      Busiest_days busiest_period_of_day max_order  \
0  Regularly busy       Normal_days       Most busy hours        10   
1  Regularly busy  Two_slowest_days    Average busy hours        10   
2  Regularly busy  Two_slowest_days       Most busy hours        10   
3      Least busy  Two_slowest_days    Average busy hours        10   
4      Least busy  Two_slowest_days       Most busy hours        10   

   loyalty_flag  average_spending  
0  New customer               NaN  
1  New customer          6.367797  
2  New customer          7.515897  
3  New customer          8.197727  
4  New customer          8.205556  

[5 rows x 22 columns]

In [17]:
# With the new column ready, we have to create a flag to denominate their spending category
ords_prods_merge.loc[ords_prods_merge['average_spending'] < 10 , 'spending_flag'] = 'Low spender'

In [18]:
# We do the same for high spender
ords_prods_merge.loc[ords_prods_merge['average_spending'] >= 10, 'spending_flag'] = 'High spender'

In [19]:
# We check the frequency
ords_prods_merge ['spending_flag'].value_counts(dropna = False)

NaN             32198650
Low spender       200858
High spender        5351
Name: spending_flag, dtype: int64

## 7. Create an order frequency flag based on the regularity of an user's ordering behaviour

In [20]:
# We create a new column, group by criteria and apply the function
ords_prods_merge['frequency_orders'] = ords_prods_merge.groupby('user_id').agg({'days_from_last_order':['median']})

In [21]:
# Check the output
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_from_last_order  first_order  product_id  add_to_cart_order  \
0                   NaN         True         196                  1   
1                  15.0        False         196                  1   
2                  21.0        False         196                  1   
3                  29.0        False         196                  1   
4                  28.0        False         196                  1   

   reordered  ... prices    price_range_loc     busiest_day      Busiest_days  \
0          0  ...    9.0  Mid-range product  Regularly busy       Normal_days   
1          1  ...    9.0  Mid-range product  Regularly busy  Two_slowest_days   
2          1  ...    9.0  Mid-range product  Regularly busy  Two_slowest_days   
3          1  ...    9.0  Mid-range product      Least busy  Two_slowest_days   
4          1  ...    9.0  Mid-range product      Least busy  Two_slowest_days   

   busiest_period_of_day max_order  loyalty_flag average_spending  \
0        Most busy hours        10  New customer              NaN   
1     Average busy hours        10  New customer         6.367797   
2        Most busy hours        10  New customer         7.515897   
3     Average busy hours        10  New customer         8.197727   
4        Most busy hours        10  New customer         8.205556   

  spending_flag  frequency_orders  
0           NaN               NaN  
1   Low spender              20.5  
2   Low spender              13.0  
3   Low spender              10.0  
4   Low spender              20.0  

[5 rows x 24 columns]

In [23]:
# With the new column ready, we have to create a flag to denominate the customers based on their spending behaviour
ords_prods_merge.loc[ords_prods_merge['frequency_orders'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [24]:
# We do the same for regular customers
ords_prods_merge.loc[(ords_prods_merge['frequency_orders'] > 10) & (ords_prods_merge ['frequency_orders'] <=20), 'frequency_flag'] = 'Regular customer'

In [25]:
# And for the frequent customers
ords_prods_merge.loc[ords_prods_merge['frequency_orders'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [26]:
# We check the results
ords_prods_merge ['frequency_flag'].value_counts(dropna = False)

NaN                      32198651
Frequent customer           86596
Regular customer            59993
Non-frequent customer       59619
Name: frequency_flag, dtype: int64

In [27]:
# To export the df as pickle
ords_prods_merge.to_pickle(os.path.join(path, '2. Data','2.2. Prepared Data', 'orders_products_merged_with_aggregations.pkl'))